File: app/src/main/java/com/nononsenseapps/feeder/model/FeedParser.java
package com.nononsenseapps.feeder.model;

import android.annotation.SuppressLint;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.Log;
import com.nononsenseapps.text.HtmlToPlainTextConverter;
import com.rometools.rome.feed.synd.SyndContent;
import com.rometools.rome.feed.synd.SyndEnclosure;
import com.rometools.rome.feed.synd.SyndEntry;
import com.rometools.rome.feed.synd.SyndFeed;
import com.rometools.rome.feed.synd.SyndLink;
import com.rometools.rome.io.SyndFeedInput;
import com.rometools.rome.io.XmlReader;
import okhttp3.Cache;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import org.joda.time.DateTime;

import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.security.cert.X509Certificate;
import java.util.List;
import java.util.concurrent.TimeUnit;

import static java.lang.Math.min;

public class FeedParser {

    // Should reuse same instance to have same cache
    private static OkHttpClient _client;

    private static OkHttpClient cachingClient(File cacheDirectory) {
        if (_client != null) {
            return _client;
        }

        int cacheSize = 10 * 1024 * 1024; // 10 MiB
        Cache cache = new Cache(cacheDirectory, cacheSize);

        OkHttpClient.Builder builder = new OkHttpClient.Builder()
                .cache(cache)
                .connectTimeout(5, TimeUnit.SECONDS)
                .readTimeout(5, TimeUnit.SECONDS);

        trustAllCerts(builder);

        _client = builder.build();

        return _client;
    }

    private static void trustAllCerts(OkHttpClient.Builder builder) {
        try {
            X509TrustManager trustManager = new X509TrustManager() {
                @SuppressLint("TrustAllX509TrustManager")
                @Override
                public void checkClientTrusted(X509Certificate[] chain, String authType) throws CertificateException {
                }

                @SuppressLint("TrustAllX509TrustManager")
                @Override
                public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
                }

                @Override
                public X509Certificate[] getAcceptedIssuers() {
                    return new X509Certificate[0];
                }
            };

            SSLContext sslContext = SSLContext.getInstance("TLS");
            sslContext.init(null, new TrustManager[]{trustManager}, null);
            SSLSocketFactory sslSocketFactory = sslContext.getSocketFactory();

            builder.sslSocketFactory(sslSocketFactory, trustManager)
                   .hostnameVerifier(((hostname, session) -> true));
        } catch (NoSuchAlgorithmException | KeyManagementException e) {
            e.printStackTrace();
        }
    }

    public static SyndFeed parseFeed(String url, File cacheDir) throws FeedParsingError {
        try {
            if (!url.startsWith("http://") && !url.startsWith("https://")) {
                url = "http://" + url;
            }

            Request request = new Request.Builder()
                    .url(url)
                    .build();

            Response response = cachingClient(cacheDir).newCall(request).execute();

            if (!response.isSuccessful()) {
                throw new IOException("Unexpected code " + response);
            }

            Log.d("RSSLOCAL", "cache response: " + response.cacheResponse());
            Log.d("RSSLOCAL", "network response: " + response.networkResponse());

            return parseFeed(response.body().source().inputStream());
        } catch (Throwable e) {
            throw new FeedParsingError(e);
        }
    }

    static SyndFeed parseFeed(InputStream is) throws FeedParsingError {
        try {
            return new SyndFeedInput().build(new XmlReader(is));
        } catch (Throwable e) {
            throw new FeedParsingError(e);
        }
    }

    @Nullable
    public static String firstEnclosure(SyndEntry entry) {
        if (!entry.getEnclosures().isEmpty()) {
            SyndEnclosure enclosure = entry.getEnclosures().get(0);
            if (enclosure.getUrl() != null) {
                return enclosure.getUrl();
            }
        }

        return null;
    }

    @Nullable
    public static String publishDate(SyndEntry entry) {
        if (entry.getPublishedDate() != null) {
            return new DateTime(entry.getPublishedDate().getTime()).toDateTimeISO().toString();
        }
        return null;
    }

    @NonNull
    public static String title(SyndEntry entry) {
        return nonNullString(entry.getTitle());
    }

    @NonNull
    public static String plainTitle(SyndEntry entry) {
        return HtmlToPlainTextConverter.HtmlToPlainText(title(entry));
    }

    @NonNull
    public static String description(SyndEntry entry) {
        // Atom
        if (!entry.getContents().isEmpty()) {
            List<SyndContent> contents = entry.getContents();
            SyndContent content = null;

            // In case of multiple contents, prioritize html
            for (SyndContent c: contents) {
                if (content == null) {
                    content = c;
                } else if ("html".equalsIgnoreCase(content.getType()) ||
                        "xhtml".equalsIgnoreCase(content.getType())) {
                    // Already html
                    break;
                } else if ("html".equalsIgnoreCase(c.getType()) ||
                        "xhtml".equalsIgnoreCase(c.getType())) {
                    content = c;
                    break;
                }
            }

            return nonNullString(content.getValue());
        }

        // Rss
        if (entry.getDescription() != null) {
            return nonNullString(entry.getDescription().getValue());
        }

        // In case of faulty feed
        return "";
    }

    /**
     *
     * @return null in case no self links exist - which is true for some RSS feeds
     */
    @Nullable
    public static String selfLink(SyndFeed feed) {
        // entry.getUri() can return bad data in case of atom feeds where it returns the ID element
        for (SyndLink link : feed.getLinks()) {
            if ("self".equalsIgnoreCase(link.getRel())) {
                return link.getHref();
            }
        }

        return null;
    }

    @NonNull
    public static String snippet(SyndEntry entry) {
        String text = HtmlToPlainTextConverter.HtmlToPlainText(FeedParser.description(entry));
        return text.substring(0, min(200, text.length()));
    }

    public static class FeedParsingError extends Exception {
        public FeedParsingError(Throwable e) {
            super(e);
        }
    }

    @NonNull
    private static String nonNullString(@Nullable String text) {
        return text == null ? "" : text;
    }
}

File: app/src/main/kotlin/com/nononsenseapps/feeder/model/FeedParser.kt
package com.nononsenseapps.feeder.model

import android.util.Log
import com.nononsenseapps.feeder.util.asFeed
import com.nononsenseapps.feeder.util.relativeLinkIntoAbsolute
import com.nononsenseapps.feeder.util.relativeLinkIntoAbsoluteOrThrow
import com.nononsenseapps.feeder.util.sloppyLinkToStrictURL
import com.nononsenseapps.jsonfeed.Feed
import com.nononsenseapps.jsonfeed.JsonFeedParser
import com.nononsenseapps.jsonfeed.cachingHttpClient
import com.nononsenseapps.jsonfeed.feedAdapter
import com.rometools.rome.io.SyndFeedInput
import com.rometools.rome.io.XmlReader
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.jsoup.Jsoup
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.net.MalformedURLException
import java.net.URL

object FeedParser {

    // Should reuse same instance to have same cache
    @Volatile private var client: OkHttpClient = cachingHttpClient()
    @Volatile private var jsonFeedParser: JsonFeedParser = JsonFeedParser(client, feedAdapter())

    /**
     * To enable caching, you need to call this method explicitly with a suitable cache directory.
     */
    @Synchronized fun setup(cacheDir: File?): FeedParser {
        this.client = cachingHttpClient(cacheDir)
        jsonFeedParser = JsonFeedParser(client, feedAdapter())

        return this
    }

    /**
     * Finds the preferred alternate link in the header of an HTML/XML document pointing to feeds.
     */
    fun findFeedUrl(html: String,
                    preferRss: Boolean = false,
                    preferAtom: Boolean = false,
                    preferJSON: Boolean = false): URL? {

        val feedLinks = getAlternateFeedLinksInHtml(html)
                .sortedBy {
                    val t = it.second.toLowerCase()
                    when {
                        preferAtom && t.contains("atom") -> "0"
                        preferRss && t.contains("rss") -> "1"
                        preferJSON && t.contains("json") -> "2"
                        else -> t
                    }
                }
                .map {
                    sloppyLinkToStrictURL(it.first) to it.second
                }

        return feedLinks.firstOrNull()?.first
    }

    /**
     * Returns all alternate links in the header of an HTML/XML document pointing to feeds.
     */
    fun getAlternateFeedLinksAtUrl(url: URL): List<Pair<String, String>> {
        val html = curl(url)
        return when {
            html != null -> getAlternateFeedLinksInHtml(html, baseUrl = url)
            else -> emptyList()
        }
    }

    /**
     * Returns all alternate links in the header of an HTML/XML document pointing to feeds.
     */
    fun getAlternateFeedLinksInHtml(html: String, baseUrl: URL? = null): List<Pair<String, String>> {
        val doc = Jsoup.parse(html.byteInputStream(), "UTF-8", "")
        val header = doc.head()

        return header.getElementsByAttributeValue("rel", "alternate")
                .filter { it.hasAttr("href") && it.hasAttr("type") }
                .filter {
                    val t = it.attr("type").toLowerCase()
                    when {
                        t.contains("application/atom") -> true
                        t.contains("application/rss") -> true
                        t.contains("application/json") -> true
                        else -> false
                    }
                }
                .filter {
                    val l = it.attr("href").toLowerCase()
                    try {
                        if (baseUrl != null) {
                            relativeLinkIntoAbsoluteOrThrow(base = baseUrl, link = l)
                        } else {
                            URL(l)
                        }
                        true
                    } catch (_: MalformedURLException) {
                        false
                    }
                }
                .map {
                    when {
                        baseUrl != null -> relativeLinkIntoAbsolute(base = baseUrl, link = it.attr("href")) to it.attr("type")
                        else -> sloppyLinkToStrictURL(it.attr("href")).toString() to it.attr("type")
                    }
                }
    }

    fun curl(url: URL): String? {
        var result: String? = null
        curlAndOnResponse(url) {
            result = body()?.string()
        }
        return result
    }

    fun curlAndOnResponse(url: URL, block: (Response.() -> Unit)) {
        val request = Request.Builder()
                .url(url)
                .build()

        val call = client.newCall(request)
        val response = call.execute()

        if (!response.isSuccessful) {
            throw IOException("Unexpected code " + response)
        }

        response.use {
            it.block()
        }
    }

    @Throws(FeedParser.FeedParsingError::class)
    fun parseFeedUrl(url: URL): Feed {
        Log.d("RxFeedParser", "parseFeed: $url")
        try {

            var result: Feed? = null
            curlAndOnResponse(url) {
                Log.d("RxRSSLOCAL", "cache response: " + cacheResponse())
                Log.d("RxRSSLOCAL", "network response: " + networkResponse())

                val isJSON = (header("content-type") ?: "").contains("json")
                val body = body()?.string()

                if (body != null) {
                    val alternateFeedLink = findFeedUrl(body, preferAtom = true)

                    val feed = if (alternateFeedLink != null) {
                        parseFeedUrl(alternateFeedLink)
                    } else {
                        when (isJSON) {
                            true -> jsonFeedParser.parseJson(body)
                            false -> parseRssAtomBody(body)
                        }
                    }

                    result = if (feed.feed_url == null) {
                        // Nice to return non-null value here
                        feed.copy(feed_url = url.toString())
                    } else {
                        feed
                    }
                } else {
                    throw NullPointerException("Response body was null")
                }
            }

            return result!!
        } catch (e: Throwable) {
            throw FeedParsingError(e)
        }

    }

    @Throws(FeedParser.FeedParsingError::class)
    fun parseRssAtomBody(feedXml: String): Feed = parseFeedInputStream(feedXml.byteInputStream())

    @Throws(FeedParser.FeedParsingError::class)
    fun parseFeedInputStream(`is`: InputStream): Feed {
        `is`.use {
            try {
                val feed = SyndFeedInput().build(XmlReader(`is`))
                return feed.asFeed()
            } catch (e: Throwable) {
                throw FeedParsingError(e)
            }
        }
    }

    class FeedParsingError(e: Throwable) : Exception(e)
}
